apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+#include "Halide.h"
+
+#ifndef user_error
+#define user_error                Halide::Internal::ErrorReport(__FILE__, __LINE__, nullptr, Halide::Internal::ErrorReport::User)
+#endif
+
+#ifndef user_warning
+#define user_warning              Halide::Internal::ErrorReport(__FILE__, __LINE__, nullptr, Halide::Internal::ErrorReport::User | Halide::Internal::ErrorReport::Warning)
+#endif
+
+#ifndef user_assert
+#define user_assert(c)            _halide_internal_assertion(c, Halide::Internal::ErrorReport::User)
+#endif
+
+#ifndef internal_assert
+#define internal_assert(c)        _halide_internal_assertion(c, 0)
+#endif
+
+#ifndef internal_error
+#define internal_error            Halide::Internal::ErrorReport(__FILE__, __LINE__, nullptr, 0)
+#endif
+
+namespace Halide {
+namespace Internal {
+namespace Autoscheduler {
+
+std::map<std::string, Box> inference_bounds(const std::vector<Function> &functions,
```

Comment:
~most of these functions aren't used outside this file, and so should be in an anonymous namespace

Reply:
Done.